### PR TITLE
vscode: added integrated support for MS VSCODE

### PR DESCRIPTION
**Pull request by are-scenic**
Added support for MS VSCODE in a form of a workspace file
(snapd.code-workspace) configured to develop snapd using this IDE,
including recommendations for the set of useful extensions.
The IDE will suggest installing these extensions when the workspace
file is opened first time.

- The test `TestDoCpErr` has been adjusted to have a workaround
  for the issue of debugger (used by VSCode Go extension) that
  incorrectly allows writing to STDIN.

Signed-off-by: Arseniy Aharonov <arseniy.aharonov@canonical.com>

### CODE REVIEW

**Comment by zyga** [comment]
I love this, it's the best gratis Go development experience.

**Comment by MiguelPires** [comment]
> The test TestDoCpErr has been adjusted to have a workaround
for the issue of gopls debugger that incorrectly allows
writing to STDIN.

Can you expand a bit on this please? Was the test failing on vscode? How does it relate to gopls (which is a language server, not a debugger)?

File: snapd.code-workspace
Comment (pedronis) [inline snapd.code-workspace:8]: as we discussed the test are not currently designed to fully work with SNAPD_DEBUG=1, not sure setting it here make sense

File: snapd.code-workspace
Comment (pedronis) [inline snapd.code-workspace:45]: same consideration

File: snapd.code-workspace
Comment (pedronis) [inline snapd.code-workspace:50]: missing final newline, it doesn't matter, but it's nicer with

**Review by pedronis — COMMENTED** [review]
some considerations

**Comment by are-scenic** [comment]
> 
> > The test TestDoCpErr has been adjusted to have a workaround
> > for the issue of gopls debugger that incorrectly allows
> > writing to STDIN.
> 
> Can you expand a bit on this please? Was the test failing on vscode? How does it relate to gopls (which is a language server, not a debugger)?

OMG! I am so sorry. I meant `delve` not `gopls`. I am sorry, lately I just tried to collect as much info about Google's Go extension for VSCode. And `gopls` got smooth integration into it lately. I can't explain for which reason this name replaced `delve` in my 
mind, but what I meant was `delve` not `gopls`. 

I'll fix the text, thank you.

**Comment by are-scenic** [comment]
Because of the embarrassing mistake in the comment of the original message (thanks @MiguelPires for catching it before it got merged), I will override the original commit with the new one (despite the general recommendation to keep the original commit that underwent a review). 
Once again, I apologize  for this mess.

File: osutil/cp_linux_test.go
Comment (MiguelPires) [inline osutil/cp_linux_test.go:41]: small nitpick: these could be asserts since, if they fail, the subsequent checks don't tell us much
```suggestion
	c.Assert(osutil.CopyFile(s.f1, s.f2, osutil.CopyFlagDefault), IsNil)
	c.Assert(os.Chmod(s.f2, 0444), IsNil)
```

File: osutil/cp_linux_test.go
Comment (MiguelPires) [inline osutil/cp_linux_test.go:50]: For simple (one liner, no error handling, etc) deferred statements, this shorthand is a bit nicer to read
```suggestion
	defer src.Close()
```

File: osutil/cp_linux_test.go
Comment (MiguelPires) [inline osutil/cp_linux_test.go:56]: Same here
```suggestion
	defer dstRo.Close()
```

**Review by MiguelPires — COMMENTED** [review]
Just a few minor improvements and nitpicks.

> I meant delve not gopls. 

Ah I see, that makes sense. Thanks for the clarification

File: snapd.code-workspace
Comment (pedronis) [inline snapd.code-workspace:49]: I am a bit unsure how we expect people to get to this comment if they hit the problem or even more before they hit the problem?

File: osutil/cp_linux_test.go
Comment (mvo5) [inline osutil/cp_linux_test.go:42]: os.Open() already opens the file "read-only" so the subsequent chmod is not needed.

**Review by pedronis — CHANGES_REQUESTED** [review]

File: osutil/cp_linux_test.go
Comment (mvo5) [inline osutil/cp_linux_test.go:40]: (nitpick) it feels a bit unclear why f1->f2 is copied here. All we need is an empty file, so `c.Assert(ioutil.WriteFile(s.f2, nil, 0444)` (if you want to keep the permissions) feels less distracting to me.

File: osutil/cp_linux_test.go
Comment (mvo5) [inline osutil/cp_linux_test.go:55]: (nitpick) I personally would not move this into a helper, given that it's only used once and is already super small. This way reading the test feels more like a "story" to me, with the function I need to jump to the function, understand it and jump back breaking my flow. But of course that is something very personal and once this is needed in multiple place extracting it would make sense.

**Review by mvo5 — COMMENTED** [review]
Thanks for working on this. I added some comments/suggestions about the test, I played a bit, I personally would do a bit more inline
```go
diff --git a/osutil/cp_linux_test.go b/osutil/cp_linux_test.go
index a7f8fa839d..9c035d9abc 100644
--- a/osutil/cp_linux_test.go
+++ b/osutil/cp_linux_test.go
@@ -20,6 +20,7 @@
 package osutil_test
 
 import (
+       "io/ioutil"
        "os"
 
        . "gopkg.in/check.v1"
@@ -36,25 +37,20 @@ func (s *cpSuite) TestCpMulti(c *C) {
        c.Check(s.f2, testutil.FileEquals, s.data)
 }
 
-func (s *cpSuite) getReadOnlyFd(c *C) *os.File {
-       c.Check(osutil.CopyFile(s.f1, s.f2, osutil.CopyFlagDefault), IsNil)
-       c.Check(os.Chmod(s.f2, 0444), IsNil)
-       f2, err := os.Open(s.f2)
-       c.Assert(err, IsNil)
-       return f2
-}
-
 func (s *cpSuite) TestDoCpErr(c *C) {
        src, err := os.Open(s.f1)
        c.Assert(err, IsNil)
-       defer func() { src.Close() }()
+       defer src.Close()
 
        st, err := src.Stat()
        c.Assert(err, IsNil)
 
-       dstRo := s.getReadOnlyFd(c)
-       defer func() { dstRo.Close() }()
+       c.Assert(ioutil.WriteFile(s.f2, nil, 0444), IsNil)
+       // Note that os.Open() opens in RO mode
+       roFd, err := os.Open(s.f2)
+       c.Assert(err, IsNil)
+       defer roFd.Close()
 
        // force an error by asking it to write to a readonly stream
-       c.Check(osutil.DoCopyFile(src, dstRo, st), NotNil)
+       c.Check(osutil.DoCopyFile(src, roFd, st), NotNil)
 }
```
but again, mostly suggestions and personal taste. Hope this is useful for you.

**Comment by pedronis** [comment]
I let other finalize the review and review the test change

**Review by mvo5 — APPROVED** [review]
The test change looks fine to me, thanks for addressing the comments. I have no opinion about the vcode part, I leave that to the others :)

File: snapd.code-workspace
Comment (zyga) [inline snapd.code-workspace:18]: This one seems like personal preference. I think it may be confusing to people.

File: snapd.code-workspace
Comment (zyga) [inline snapd.code-workspace:19]: This one is also something I would not put into the workspace. All the users who want vim or emacs behavior in vscode already have that enabled.

File: snapd.code-workspace
Comment (zyga) [inline snapd.code-workspace:15]: This one seems like a personal preference and is unrelated to Go or snapd.

**Review by zyga — COMMENTED** [review]
Mostly this is a nice improvement. My only comment is about the set of recommended packages. I think that section should be conservative. Personally I would only put the go extensions and remove anything that is personal preference of the submitter (vim or emacs keymaps and things like that).

**Review by MiguelPires — COMMENTED** [review]
The unit test LGTM. I installed vscode to see the effect of the workspace file and I agree with Zygmunt. Especially since it shows a prompt asking "Do you want to install the recommended plugins?" to which you can click yes. I think it makes sense to only have the ones that are always appropriate (Go, C, and possibly git) so you can click yes and be ready to work. But I'm not a regular user so take my opinion with a grain of salt.

**Comment by are-scenic** [comment]
> The unit test LGTM. I installed vscode to see the effect of the workspace file and I agree with Zygmunt. Especially since it shows a prompt asking "Do you want to install the recommended plugins?" to which you can click yes. I think it makes sense to only have the ones that are always appropriate (Go, C, and possibly git) so you can click yes and be ready to work. But I'm not a regular user so take my opinion with a grain of salt.

File: snapd.code-workspace
Comment (zyga) [reply snapd.code-workspace:19]: I don't think that applies. People who don't use vscode probably don't want vim and emacs modes installed as well. Things like this belong in a tutorial or in personal preference file.

File: snapd.code-workspace
Comment (zyga) [reply snapd.code-workspace:15]: So I think we just disagree on expectations. I was expecting minimal support for using vscode with snapd. Here the Go extension is pretty much the only thing that is absolutely required and this is what I would expect to use.

We can leave recommendations and comments but the key difference is that the IDE won't prompt the (possibly inexperienced) user to install those.

File: snapd.code-workspace
Comment (are-scenic) [reply snapd.code-workspace:19]: @zyga I see your point. I thought actually the other way around. I thought that vim emacs users my want to use VSCODE just for debugging and I wanted to make them feel home. But perhaps I am trying to much. "The only enemy of good is perfect."

File: snapd.code-workspace
Comment (anonymouse64) [reply snapd.code-workspace]: `-check.vv` only outputs stuff in the case where tests pass if you use `-v` too, but `-check.vv` should still output test names, etc. when they fail without `-v`. I can see the utility of seeing tests run in real time as opposed to waiting until they are all done executing to see the failures

**Comment by anonymouse64** [comment]
I agree with others, the only extension we should suggest here is the Go one `golang.go` is the extension ID. Everything else is personal preference

**Comment by are-scenic** [comment]
@zyga, @MiguelPires and @anonymouse64, guys thank you all for your elaborate inputs.  

I understand your point and I would love to suggest the following:
1. Extension `golang.go` - we all agree that his one is relevant
2. Extension `766b.go-outliner`:
  -`golang.go` is weak at outlining contents of a go-package that spreads over several files, which we do a lot;  
  -So, `766b.go-outliner` comes to save the day;  
  -I strongly recommend to keep this one too
3. Extension `ms-vscode.cpptools-extension-pack` - I agree with Miguel that this one is important too, since we do have C
4. Extension `jerrygoyal.shortcut-menu-bar` - in my opinion is a must, and I really don't understand why it's not part of the UI  
  -It adds to UI a small toolbar of buttons without which UI is really crippled (e.g. "Back" and "Forward")
  -I vote to keep it too
5. All the rest - I will comment them out and will add expressive comments explaining who might benefit from those.  
  -I believe this is important, since today VSCode library of extensions is so bloated that it is really hard to find the best one among the suggested enormous amount of extensions for the same thing  

Please, provide your opinion (please everybody), so that we can decide on what is best.

Thank you,
Arseniy

**Comment by anonymouse64** [comment]
I think there's a fair point about including the official C/C++ extension from Microsoft, the rest should all not be included IMHO. We don't need to render opinions on the default usability of VS code here, I'm fine if you want to maintain your own set of recommended extensions and leave a link to that recommendation somewhere (perhaps in a comment) but I don't think we want it in the default set of recommended extensions.

**Comment by anonymouse64** [comment]
Also, the go-outline extension has not received an update since 2018 as per the changelog which leads me to believe that it is unmaintained

**Comment by MiguelPires** [comment]
I agree with Ian. The plugins for Go and C/C++ are fair since they're "official" and used for languages that snapd uses. The rest are based on personal opinions of the vscode UI which might be very valid but they're opinions nonetheless. Also, googling "vscode golang outline packages" turns up 766b.go-outliner as the first hit so devs frustrated with the vscode UI will probably handle it on their own.

**Comment by are-scenic** [comment]
I think we need to label this one with `skip_spread` label.

**Comment by are-scenic** [comment]
I've made the changes following our discussion. Please review.

**Comment by anonymouse64** [comment]
need to close and re-open this for skip-spread to work unfortunately

**Review by anonymouse64 — COMMENTED** [review]
all of @MiguelPires suggestion look good to me, with those suggestions applied +1 from me

**Comment by are-scenic** [comment]
> all of @MiguelPires suggestion look good to me, with those suggestions applied +1 from me

I totally agree with you, all of the suggestions of @MiguelPires make these comments objective and informative.
Comments like the code itself shell be clean and essential. As I read it now, my comments were really terrible.

@MiguelPires, thank you. :-)

**Review by MiguelPires — APPROVED** [review]
Thanks

**Pull request merged**
